### ISSUE TITLE
best way to update OS X Yosemite after using CAT

### ISSUE THREAD

**Issue by EricPSmith** [dokterdok/Continuity-Activation-Tool#139]
Hi folks,

Just curious whether there are any precautions necessary when updating OS X Yosemite (e.g. to 10.10.2) after using CAT to get Continuity working on my Mac. I'm using a GMYLE dongle on a late 2011 MacBook Pro, in case that matters. 

I'm especially wondering whether the modified kexts in /System could impact the update in any way, and whether I can safely assume that the OS X update will simply overwrite the modified versions with clean, updated versions. 

I'm also wondering what happens to the backed up kexts in my Documents folder, like whether they are simply overwritten with new versions when CAT is run again after the update, and whether I should be keeping copies of the pre-update versions for any reason. 

And I'm wondering whether CAT would ever need an update to handle a given OS X update. 

It would be easy enough to undo CAT before the update, and then redo CAT after the update, but it would also be nice to skip the extra step if all of this is cleanly handled for me, especially since there are likely more OS X Yosemite updates coming. 

Thanks for any info you can provide.

Eric

**Comment by dokterdok** [comment]
Hi @EricPSmith,
Good questions, I'm answering them separately.
**TL;DR**: If an OS X update breaks Continuity, just apply CAT again, it's safe.

>I'm especially wondering whether the modified kexts in /System could impact the update in any way, and whether I can safely assume that the OS X update will simply overwrite the modified versions with clean, updated versions.

That's correct, when Apple provides an OS X update with fixes to the wireless drivers, it overwrites the existing ones with clean, updated ones. When there aren't Wi-Fi or Bluetooth related fixes in an OS X update, then the existing patched kexts will stay as such and won't need to be re-patched by CAT.

>I'm also wondering what happens to the backed up kexts in my Documents folder, like whether they are simply overwritten with new versions when CAT is run again after the update, 

If CAT detects existing backed up kexts, it will ask you whether you'd like to overwrite the existing backups or not during the activation procedure.

>and whether I should be keeping copies of the pre-update versions for any reason.

It's useful to keep a backup of your previously patched kexts, in case CAT doesn't know how to patch the kexts that come with an OS X update. That hasn't happened so far, but in that scenario, your best bet is to

1. make a manual backup of the new OS X IOBluetoothFamily.kext and IO80211Family.kext (located in /System/Library/Extensions).
2. [manually install](http://www.cindori.org/software/kextdrop) older kexts that are known to work, e.g. the ones that are stored by CAT in the KextsBackupAfterPatch folder.
3. see if your system is stable or not. As far as I know, OS X has not been designed to run with a mix of old/new kext versions and might become unstable.

> And I'm wondering whether CAT would ever need an update to handle a given OS X update.
Maybe, we don't know.

CAT has been built to detect the patterns it needs to modify in the kexts before patching, independently from the OS X kext version. If it doesn't find the patterns, then the activation will abort and no changes are made. So far, it has worked with all OS X versions, because the patterns that need to be modified have a small footprint in the kexts and can be easily isolated. So we can just hope CAT continues to be compatible with future OS X versions. In the worst case, it's worth restoring and testing an old kext backup as described above.

>It would be easy enough to undo CAT before the update, and then redo CAT after the update, but it would also be nice to skip the extra step if all of this is cleanly handled for me, especially since there are likely more OS X Yosemite updates coming.

CAT doesn't need to be uninstalled before an OS X update. The worst case scenario is that the patches are undone by clean new kexts replacing the old patched kexts. You'll just get regular Wi-Fi / Bluetooth but no Handoff / New AirDrop etc. To activate again, just use the CAT activation. An automatic update could technically could be a nice addition, I've written a draft of how it could work [here](http://apple.stackexchange.com/questions/163667/automate-continuity-reactivation-post-update/163836#163836). I've no plans to implement this, mainly due to the related risks of automating low-level mods with unknown, future OS X releases. The effort/benefit ratio is also not worth it IMO.

**Comment by EricPSmith** [comment]
Hi dokterdok,

Thank you for the clear explanation. The interaction between CAT and OS X makes much more sense to me now. 

I went ahead with the Yosemite 10.10.2 update, using what I thought was a fairly cautious approach, yet ran into CAT-related troubles even so. 

Going into this, I had new AirDrop, as well as my Apple wireless keyboard & trackpad, working fine with my MacBook Pro late 2011 (OS X Yosemite 10.10.1) through a GMYLE Bluetooth LE dongle, activated via CAT 2.0.0.

I started the OS X Yosemite 10.10.2 update by quitting all apps, verifying both my internal drive and the startup volume on it with disk utility, and then repairing disk permissions. I then proceeded with installing the OS X update through the Mac App Store, leaving my Bluetooth on and my wireless keyboard & mouse in use during the update. In retrospect, leaving Bluetooth on may have been behind at least some of my troubles. 

Upon completion of the OS X update, I was surprised to find the light on my Bluetooth dongle flashing (i.e. in use), as I had expected the update to force a return to the built-in Bluetooth hardware. (The dongle’s light goes on solid when powered but without a driver.) At this point my wireless keyboard and mouse seemed to be working fine, but it’s not clear to me which Bluetooth hardware they were using. When I checked for AirDrop, the Mac did not see my iPhone, and my iPhone did not see my Mac. 

I then ran CAT 2.0.0 to reactivate continuity, which appeared to run fine, and reported no issues. Following the requisite restart, my wireless trackpad was extremely jumpy and erratic, to the point of being unusable. Unpairing and repairing seemed to help very briefly, but then the completely erratic behavior returned. Meanwhile, my wireless keyboard was mostly working, but occasionally showed mysterious behavior, like sporadically changing the OS X volume setting (as seen by HUD) without my ever getting near the top row of keys. (I’ve used this keyboard for nearly three years, and have never seen this behavior. I couldn’t figure out how to reliably reproduce it.)

Seeing that my system was confused, I turned off Bluetooth and then uninstalled CAT. 

After the requisite restart, I encountered even more bizarre behavior. With the OS X Bluetooth menu indicating that Bluetooth was off, and with the Bluetooth dongle’s light on solid, my wireless keyboard was still working. 

I turned Bluetooth back on after the CAT uninstall, and my wireless trackpad started working, but with the same erratic behavior as before. Meanwhile, in the Bluetooth system preferences pane, my wireless keyboard status was ‘not connected,’ yet it continued to work. 

At this point, I removed both the keyboard and trackpad from the Bluetooth paired devices list. Somehow, my wireless keyboard still worked. 

Finally, with Bluetooth left on, I shut down my Mac, removed the Bluetooth dongle, and then powered back up. Once the Mac powered up, my wireless keyboard & trackpad went back to working normally, with no further issues. 

—————

After taking a break from all the disruption, and confirming over days that my native Bluetooth was working properly in 10.10.2, I gave CAT another shot. 

I saw on the CAT beta page a 2.0.1 release, but it was not clear to me how to download that version in executable form. Not knowing which version I would get, I redownloaded CAT using the Download Zip button and replaced the copy I had with that fresh one. When launched, it indicated v2.0.0. 

With my Mac running fine on OS X 10.10.2 with built-in Bluetooth, I quit all apps, turned off Bluetooth as a preventive measure, and ran CAT’s activate continuity command. The script aborted, complaining of no Bluetooth 4.0 hardware, with no prompt to insert dongle (I had gotten such a prompt with 10.10.1). 

I returned to the CAT main menu, inserted the Bluetooth dongle, and ran activate continuity again. The script ran through to completion this time without issue, other than overwriting backup kexts without asking for permission first or offering an alternative. 

Before proceeding with the script-prompted reboot, I used Disk Utility to repair permissions on the startup volume, which fixed permissions on a bunch of items in the two patched kexts in /System/Library/Extensions. Then I rebooted. 

Following the restart, the Bluetooth dongle’s light was solid, and the Bluetooth menu still indicated that Bluetooth was off, yet my wireless keyboard was somehow working even so (but not my wireless trackpad). 

I went to restart again to see if this strange behavior would clear up, and inexplicably during the shutdown process got a password prompt from CAT saying it wanted to make changes. I never give OS X permission to reopen apps on restart, so I don’t know how CAT could have been running after its own terminating reboot. Canceling this password prompt allowed CAT to quit and the restart to proceed. 

After this second restart, my wireless keyboard continued to work despite the Bluetooth menu’s indicating that Bluetooth was off, and I went ahead anyway. Turning Bluetooth on caused the dongle’s light to go from solid to blinking; at the same time, Pair buttons appeared beside my wireless keyboard and my wireless trackpad in the Bluetooth system preferences device list. I was able to pair each device successfully, and so far both are working well. 

I tested AirDrop from my iPhone to my Mac, and it worked perfectly on the first try. When I tried the other direction, the Mac saw the iPhone, but got stuck Waiting… once the transfer was initiated. As I found before, briefly enabling Airplane Mode on iPhone cleared up the block, and the file transfer completed successfully. 

I'm really happy and grateful to have AirDrop and the other Continuity features again. At the same time, I'm a bit disappointed by how messy this process has been, preserving those features through an OS X update. Perhaps these detailed notes will shed light on how to make it smoother and easier. 

Many thanks,

Eric
